draft-ietf-quic-qpack.md

```diff
@@ -101,19 +101,238 @@ balance between resilience against head-of-line blocking and optimal compression
+permitted by the value of SETTINGS_QPACK_BLOCKED_STREAMS, compression efficiency
+can be improved by referencing dynamic table entries that are still in transit,
+but if there is loss or reordering the stream can become blocked at the decoder.
+An encoder avoids the risk of blocking by only referencing dynamic table entries
+which have been acknowledged, but this means using literals. Since literals make
+the header block larger, this can result in the encoder becoming blocked on
+congestion or flow control limits.
+
+An encoder MUST limit the number of streams which could become blocked to the
+value of SETTINGS_QPACK_BLOCKED_STREAMS at all times. Note that the decoder
+might not actually become blocked on every stream which risks becoming blocked.
+If the decoder encounters more blocked streams than it promised to support, it
+MUST treat this as a stream error of type HTTP_QPACK_DECOMPRESSION_FAILED.
+
+### Largest Known Received
+
+For the encoder to identify which dynamic table entries can be safely used
+without a stream becoming blocked, the encoder tracks the absolute index of the
+decoder's Largest Known Received entry.
+
+When blocking references are permitted, the encoder uses acknowledgement of
+header blocks to identify the Largest Known Received index, as described in
+{{header-acknowledgement}}.
+
+To acknowledge dynamic table entries which are not referenced by header blocks,
+for example because the encoder or the decoder have chosen not to risk blocked
+streams, the decoder sends a Table State Synchronize instruction (see
+{{table-state-synchronize}}).
+
+
+### Speculative table updates {#speculative-updates}
+
+Implementations can *speculatively* send instructions on the encoder stream
+which are not needed for any current HTTP request or response.  Such headers
+could be used strategically to improve performance.  For instance, the encoder
+might decide to *refresh* by sending Duplicate representations for popular
+header fields ({{duplicate}}), ensuring they have small indices and hence
+minimal size on the wire.
+
+## Decoder
+
+<!-- new -->
+
+Like in HPACK, the decoder processes header blocks and emits the corresponding
+header lists. It also processes dynamic table modifications from instructions on
+the encoder stream.
+
+A decoder MUST must emit header fields in the order their representations appear
+in the input header block.
+
+
+### State Synchronization
+
+The decoder stream ({{decoder-stream}}) signals key events at the decoder that
+permit the encoder to track the decoder's state.  These events are:
+
+- Complete processing of a header block
+- Abandonment of a stream which might have remaining header blocks
+- Receipt of new dynamic table entries
+
+Regardless of whether a header block contained blocking references, the
+knowledge that it has been processed permits the encoder to evict
+entries to which no unacknowledged references remain; see {{blocked-insertion}}.
```

Comment:
Now that we only acknowledge blocks with dynamic references, perhaps this should read "Knowledge that a header block which contains references to the dynamic table has been processed permits [...], regardless of whether those references were potentially blocking."